apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/signature/SignatureParser.java

```diff
@@ -32,10 +36,60 @@
 
 public class SignatureParser {
 
+    /**
+     * If the cache reaches this size we assume that the application creates a lot of dynamic queries.
+     * In that case it's inefficient to try to cache these as they are not likely to be repeated.
+     * But we still pay the price of allocating a Map.Entry and a String for the signature.
+     */
+    private static final int DISABLE_CACHE_THRESHOLD = 512;
+    /**
+     * The cache management overhead is probably not worth it for short queries
+     */
+    private static final int QUERY_LENGTH_CACHE_LOWER_THRESHOLD = 64;
+    /**
+     * We don't want to keep alive references to huge query strings
+     */
+    private static final int QUERY_LENGTH_CACHE_UPPER_THRESHOLD = 10_000;
     private final Scanner scanner = new Scanner();
+    /**
+     * Not using weak keys as the DISABLE_CACHE_THRESHOLD may not be reached when clearing stale entries.
+     * That means that we are leaking Strings but as the size of the map is limited that should not be an issue.
+     * Another reason is that ORMs like Hibernate generate equal SQL strings for the same query but don't reuse the same string instance.
+     * When relying on weak keys, we would not leverage any caching benefits if the query string is collected.
```

Comment:
That what I meant by saying you should count yourself (so - not to count you as a PreparedStatement 😛 ), eg count removals. But I assume if we have one cache for the entire JVM this is a non-issue